### PR TITLE
Feature: Add rotation module feature

### PR DESCRIPTION
**Pull request by xhycode**

### CODE REVIEW

File: Marlin/src/module/configuration_store.cpp [hunk 1]
```diff
@@ -127,10 +127,10 @@ typedef struct SettingsDataStruct {
 
   planner_settings_t planner_settings;
```
Comment (snapmaker-lixiaolong) [inline Marlin/src/module/configuration_store.cpp:128]: you forgot to update the version

**Review by snapmaker-lixiaolong — APPROVED** [review]

File: Marlin/src/module/planner.cpp [hunk 2]
```diff
@@ -2340,15 +2348,15 @@ bool Planner::_populate_block(block_t * const block, bool split_move,
...
+      LIMIT_ACCEL_LONG(Z_AXIS, 0);
       LIMIT_ACCEL_LONG(E_AXIS, ACCEL_IDX);
     }
     else {
-      LIMIT_ACCEL_FLOAT(A_AXIS, 0);
-      LIMIT_ACCEL_FLOAT(B_AXIS, 0);
-      LIMIT_ACCEL_FLOAT(C_AXIS, 0);
+      LIMIT_ACCEL_FLOAT(X_AXIS, 0);
+      LIMIT_ACCEL_FLOAT(Y_AXIS, 0);
+      LIMIT_ACCEL_FLOAT(Z_AXIS, 0);
```
Comment (scotthsl) [inline Marlin/src/module/planner.cpp:2359]: Why don't you limit acceleration of B axis?

File: config/default/Configuration.h [hunk 3]
```diff
@@ -982,6 +982,7 @@
 #define X_HOME_DIR -1
 #define Y_HOME_DIR -1
 #define Z_HOME_DIR -1
+#define B_HOME_DIR -1
```
Comment (scotthsl) [inline config/default/Configuration.h:985]: No need to change this file.

File: config/default/Configuration_adv.h [hunk 4]
```diff
@@ -454,7 +454,8 @@
 #define X_HOME_BUMP_MM 5
 #define Y_HOME_BUMP_MM 5
 #define Z_HOME_BUMP_MM 2
-#define HOMING_BUMP_DIVISOR { 2, 2, 4 }  // Re-Bump Speed Divisor (Divides the Homing Feedrate)
+#define B_HOME_BUMP_MM 0
```
Comment (scotthsl) [inline config/default/Configuration_adv.h:457]: no need to change this file

File: snapmaker/src/module/rotary_module.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * Detection of module.
+     */
+    // void Detect();
+
+    /**
+     * Check if module is activated.
+     */
+    ROTATE_STATE_E Status() {return status_;}
+    void Status(ROTATE_STATE_E s) {status_ = s;}
```
Comment (scotthsl) [inline snapmaker/src/module/rotary_module.h:58]: Function name to set/get property should not begin with a capital letter.

File: snapmaker/src/module/toolhead_laser.cpp [hunk 6]
```diff
@@ -234,10 +235,15 @@ void ToolHeadLaser::TryCloseFan() {
 
 ErrCode ToolHeadLaser::LoadFocus() {
   CanStdMesgCmd_t cmd;
-
+  uint8_t Data[1];
```
Comment (scotthsl) [inline snapmaker/src/module/toolhead_laser.cpp:238]: Variable name should not begin with a capital letter.

File: snapmaker/src/service/system.h [hunk 7]
```diff
@@ -195,8 +195,9 @@ typedef struct {
 
   uint16_t  feedrate;    // mm/min
 
-  uint32_t  laser_power; // (uint32_t)(float * 1000)
-  uint32_t  cnc_rpm;
+  uint32_t  laser_power_cnc_rpm; // laser_power (uint32_t)(float * 1000)
+ 
+  uint32_t b;
```
Comment (scotthsl) [inline snapmaker/src/service/system.h:200]: Please add comment to describe the 'b', or give it a better name

**Review by scotthsl — CHANGES_REQUESTED** [review]
see comment for details

File: snapmaker/src/module/toolhead_laser.cpp [hunk 8]
```diff
@@ -234,10 +235,15 @@ void ToolHeadLaser::TryCloseFan() {
...
+  uint8_t data[1];
+  if (rotaryModule.status() != ROTATE_OFFLINE) {
+    data[0] = 1;
+  } else {
+    data[0] = 0;
+  }
   cmd.id     = msg_id_get_focus_;
-  cmd.length = 0;
-
+  cmd.length = 1;
```
Comment (snapmaker-lixiaolong) [inline snapmaker/src/module/toolhead_laser.cpp:245]: useing sizeof data would be better

File: snapmaker/src/module/toolhead_laser.cpp [hunk 9]
```diff
@@ -294,15 +300,21 @@ ErrCode ToolHeadLaser::SetFocus(SSTP_Event_t &event) {
...
+  if (rotaryModule.status() != ROTATE_OFFLINE) {
+    buff[2] = 1;
+  } else {
+    buff[2] = 0;
+  }
 
   cmd.id     = MODULE_FUNC_SET_LASER_FOCUS;
   cmd.data   = buff;
-  cmd.length = 2;
+  cmd.length = 3;
```
Comment (snapmaker-lixiaolong) [inline snapmaker/src/module/toolhead_laser.cpp:311]: useing sizeof data would be better

**Review by snapmaker-lixiaolong — APPROVED** [review]

**Review by scotthsl — APPROVED** [review]

File: snapmaker/src/module/toolhead_laser.cpp (hunk 8 above)
Comment (scotthsl) [reply snapmaker/src/module/toolhead_laser.cpp:245]: I don't think so, sometimes the length of buffer array will be larger then the used length

File: snapmaker/src/module/toolhead_laser.cpp (hunk 9 above)
Comment (scotthsl) [reply snapmaker/src/module/toolhead_laser.cpp:311]: I don't think so, sometimes the length of buffer array will be larger then the used length

**Pull request merged**